File: 2020/3.cpp
#include <bits/stdc++.h>
#define int long long int

using namespace std;

int slope(vector<string> input, int rt, int dwn) {
  int ROW_LEN = input[0].length();

  int trees = 0;
  for(int i = dwn, j = rt; i < input.size(); i += dwn) {
    if(input[i][j] == '#') trees++;
    j += rt;
    j %= ROW_LEN;
  }

  return trees;
}

signed main() {
  freopen("3.txt", "r", stdin);

  vector<string> input;
  string l;
  while(cin >> l) {
    input.push_back(l);
  }

  int two = slope(input, 1, 1) * slope(input, 3, 1) *
            slope(input, 5, 1) * slope(input, 7, 1) *
            slope(input, 1, 2);
  
  cout << slope(input, 3, 1) << endl;
  cout << two << endl;
}

File: 2020/3.js
const fs = require("fs");
const input = fs
  .readFileSync("./3.txt")
  .toString()
  .split("\n")
  .filter((x) => !!x)
  .map((x) => x.split(""));

const tree = (x) => x === "#";
const open = (x) => x === ".";
const ROW_LEN = input[0].length;

function one(input) {
  return slope(input, 3, 1);
}

function slope(input, right, down) {
  let trees = 0;
  for (let i = down, j = right; i < input.length; i += down) {
    if (tree(input[i][j % ROW_LEN])) trees++;
    j += right;
    j %= ROW_LEN;
  }

  return trees;
}

function stringify(input) {
  return input.map((x) => x.join("")).join("\n");
}

function paint(input, right, down) {
  input = [...input];
  for (let i = down; i < input.length; i += down) {
    const y = i,
      x = (i * right) % ROW_LEN;
    tree(input[y][x]) ? (input[y][x] = "X") : (input[y][x] = "O");
  }

  return stringify(input);
}

function two(input) {
  return (
    slope(input, 1, 1) *
    slope(input, 3, 1) *
    slope(input, 5, 1) *
    slope(input, 7, 1) *
    slope(input, 1, 2)
  );
}

console.log(one(input));
console.log(two(input));
